src/clj/puppetlabs/http/client/async.clj

```diff
@@ -326,39 +137,81 @@
     (.start client)
     client))
 
-(schema/defn execute-with-consumer
-  [client :- HttpAsyncClient
-   request :- HttpRequest
-   future-callback :- FutureCallback
-   result :- common/ResponsePromise
-   opts :- common/RequestOptions
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Map the Ring request onto the Java API
+
+(schema/defn callback-response :- common/Response
+  [opts :- common/UserRequestOptions
    callback :- common/ResponseCallbackFn
-   http-context :- HttpContext]
-  (let [;; Create an Apache AsyncResponseConsumer that will return the response to us as soon as it is 
-        ;; available then send the response body asynchronously
-        consumer (StreamingAsyncResponseConsumer.
-                  (FnDeliverable.
-                   (fn
-                     [http-response]
-                     (complete-response result opts callback http-response http-context))))
-        ;; If an error occurs early in the request, the consumer may not get a chance to return the
-        ;; response using the FnDeliverable. This wrapper around the future-callback guarantees that
-        ;; happens. It also takes care of notifying the consumer of the final result.
-        streaming-complete-callback (reify
-                                      FutureCallback
-                                      (completed [this http-response]
-                                        (.setFinalResult consumer nil)
-                                        (.completed future-callback http-response))
-                                      (failed [this ex]
-                                        (.setFinalResult consumer ex)
-                                        (.failed future-callback ex))
-                                      (cancelled [this]
-                                        (.setFinalResult consumer nil)
-                                        (.cancelled future-callback)))]
-    (.execute client
-              (HttpAsyncMethods/create request)
-              consumer
-              streaming-complete-callback)))
+   response :- common/Response]
+  (if callback
+    (try
+      (callback response)
+      (catch Exception e
+        {:opts opts :error e}))
+    response))
+
+(schema/defn java-content-type->clj :- common/ContentType
+  [java-content-type]
+  (if java-content-type
```

Comment:
It appears that this implementation would cause a response with an HTTP header of "Content-Type=" (note no value to the right of the "=" sign) to throw an exception back to the caller.  I got this when I tried:

~~~~
ParseException Invalid content type:   org.apache.http.entity.ContentType.parse (ContentType.java:273)
~~~~

The old implementation does an `if (empty? ...` check on the content-type before trying to use it and, as a result, delivers a Content-Type of `nil` with an otherwise full response body.

Arguably, I suppose, this kind of header value might be considered malformed.  Since it's possible for a server to do this, though, I think it would be good to preserve the spirit of the error handling that the old implementation did for this case.